core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java

```diff
@@ -0,0 +1,196 @@
+    public void handleNameResolutionError(Status error) {
+      getDelegate().handleNameResolutionError(error);
+    }
+
+    @Override
+    public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+      getDelegate().handleSubchannelState(subchannel, stateInfo);
+    }
+
+    @Override
+    public void shutdown() {
+      getDelegate().shutdown();
+      setDelegate(null);
+    }
+
+    @VisibleForTesting
+    LoadBalancer getDelegate() {
+      return delegate;
+    }
+
+    @VisibleForTesting
+    void setDelegate(LoadBalancer delegate) {
+      this.delegate = delegate;
+    }
+
+    @VisibleForTesting
+    LoadBalancer.Factory getDelegateFactory() {
+      return delegateFactory;
+    }
+
+    @VisibleForTesting
+    void setDelegateFactory(LoadBalancer.Factory delegateFactory) {
+      this.delegateFactory = delegateFactory;
+    }
+
+    /**
+     * Picks a load balancer based on given criteria.  In order of preference:
+     *
+     * <ol>
+     *   <li>User provided lb on the channel.  This is a degenerate case and not handled here.</li>
+     *   <li>gRPCLB if on the class path and any gRPC LB balancer addresses are present</li>
+     *   <li>RoundRobin if on the class path and picked by the service config</li>
+     *   <li>PickFirst if the service config choice does not specify</li>
+     * </ol>
+     *
+     * @param servers The list of servers reported
+     * @param config the service config object
+     * @return the new load balancer factory, or null if the existing lb should be used.
+     */
+    @Nullable
+    @VisibleForTesting
+    static LoadBalancer.Factory decideLoadBalancerFactory(
+        List<EquivalentAddressGroup> servers, Map<String, Object> config) {
+
+
+      // Check for balancer addresses
+      boolean haveBalancerAddress = false;
+      for (EquivalentAddressGroup s : servers) {
+        if (s.getAttributes().get(GrpcAttributes.ATTR_LB_ADDR_AUTHORITY) != null) {
+          haveBalancerAddress = true;
+          break;
+        }
+      }
+
+      if (haveBalancerAddress) {
+        try {
+          Class<?> lbFactoryClass;
+          lbFactoryClass = Class.forName(GRPCLB_LOAD_BALANCER_FACTORY_NAME);
+          Method getInstance = lbFactoryClass.getMethod("getInstance");
+          return (LoadBalancer.Factory) getInstance.invoke(null);
+        } catch (RuntimeException e) {
+          throw e;
+        } catch (Exception e) {
+          throw new RuntimeException("Can't get GRPCLB, but balancer addresses were present", e);
+        }
+      }
+
+      String serviceConfigChoiceBalancingPolicy =
+          ServiceConfigUtil.getLoadBalancingPolicy(config);
+
+      // Check for an explicitly present lb choice
+      if (serviceConfigChoiceBalancingPolicy != null) {
+        if (serviceConfigChoiceBalancingPolicy.toUpperCase(Locale.ROOT).equals("ROUND_ROBIN")) {
+          ClassNotFoundException caught = null;
+          try {
+            Class<?> lbFactoryClass;
```

Comment:
unnecessary separation of declaration and assignment.

Reply:
Done.